File: solver/oc_solver.py
"""
OC 求解器 — `求解器模块/OC_solver.m` 的 Python (numpy) 移植

与 MATLAB 版本逐行语义一致：
  - 可配置的数值参数和单步密度变化上限（xmin / move）
  - 被动空域 / 被动实体域 / active_mask
  - 体积约束的可行性检查（含舍入容差截断）
  - 拉格朗日乘子二分搜索 + info 诊断信息
"""

from __future__ import annotations

import numpy as np


def _set_default(opts: dict, name: str, value):
    if name not in opts or opts[name] is None:
        opts[name] = value
    return opts


def _make_mask(value, expected_size, default_value) -> np.ndarray:
    """构建逻辑掩码。空值 → 全 default_value；否则校验形状与合法性。"""
    if value is None or len(np.asarray(value)) == 0:
        return np.full(expected_size, default_value, dtype=bool)
    value = np.asarray(value)
    if value.shape != expected_size:
        raise ValueError(
            f"掩码形状应为 {expected_size}，实际 {value.shape}")
    if value.dtype != bool and not np.all(np.isfinite(value)):
        raise ValueError("掩码必须是逻辑数组或有限数值数组")
    return value.astype(bool)


def oc_update(x: np.ndarray, dc: np.ndarray, volfrac: float,
              opts: dict = None) -> tuple[np.ndarray, dict]:
    """增强 OC 密度更新器（Python 移植 OC_solver.m）。

    参数:
        x:       当前密度场，任意形状（如 (nely, nelx)）
        dc:      滤波后的柔顺度灵敏度，形状与 x 一致
        volfrac: 体积分数目标，0~1
        opts:    可选字段
            xmin           密度下限，默认 1e-3
            move           每步最大密度变化，默认 0.2
            tol_lambda     拉格朗日乘子二分精度，默认 1e-4
            max_bisect     最大二分次数，默认 100
            passive_void   被动空域掩码，始终为 xmin
            passive_solid  被动实体掩码，始终为 1
            active_mask    可设计单元掩码

    返回:
        (xnew, info)
        info: {lambda, bisect_iterations, volume, volume_fraction,
               volume_error, active_elements, converged}
    """
    opts = dict(opts or {})
    _set_default(opts, "xmin", 1e-3)
    _set_default(opts, "move", 0.2)
    _set_default(opts, "tol_lambda", 1e-4)
    _set_default(opts, "max_bisect", 100)
    _set_default(opts, "passive_void", None)
    _set_default(opts, "passive_solid", None)
    _set_default(opts, "active_mask", None)
    _set_default(opts, "volume_sensitivity", None)
    _set_default(opts, "volume_fn", None)

    x = np.asarray(x, dtype=float)
    dc = np.asarray(dc, dtype=float)
    if not np.all(np.isfinite(x)) or x.size == 0:
        raise ValueError("x 必须为有限非空数值")
    if dc.shape != x.shape or not np.all(np.isfinite(dc)):
        raise ValueError("dc 必须为有限数值且与 x 同形状")
    if not (0 <= volfrac <= 1):
        raise ValueError(f"volfrac 必须位于 [0,1]: {volfrac}")

    xmin = float(opts["xmin"])
    move = float(opts["move"])
    tol_lambda = float(opts["tol_lambda"])
    max_bisect = int(opts["max_bisect"])
    if not (0 < xmin < 1):
        raise ValueError(f"opts.xmin 必须位于 (0,1): {xmin}")
    if move <= 0:
        raise ValueError(f"opts.move 必须为正: {move}")
    if tol_lambda <= 0:
        raise ValueError(f"opts.tol_lambda 必须为正: {tol_lambda}")
    if max_bisect <= 0:
        raise ValueError(f"opts.max_bisect 必须为正整数: {max_bisect}")

    n = x.size
    passive_void = _make_mask(opts["passive_void"], x.shape, False)
    passive_solid = _make_mask(opts["passive_solid"], x.shape, False)
    if np.any(passive_void & passive_solid):
        raise ValueError("passive_void 与 passive_solid 不得重叠")

    if opts["active_mask"] is None:
        active_mask = ~(passive_void | passive_solid)
    else:
        active_mask = _make_mask(opts["active_mask"], x.shape, True)
        if np.any(active_mask & (passive_void | passive_solid)):
            raise ValueError("单元不能同时为可设计与被动")

    # 未被显式标记为可设计的单元固定在当前密度
    fixed_mask = ~active_mask
    fixed_values = np.minimum(1.0, np.maximum(xmin, x))
    fixed_values[passive_void] = xmin
    fixed_values[passive_solid] = 1.0

    target_volume = volfrac * n
    fixed_volume = float(fixed_values[fixed_mask].sum())
    n_active = int(np.count_nonzero(active_mask))
    min_volume = fixed_volume + n_active * xmin
    max_volume = fixed_volume + n_active
    volume_tolerance = max(tol_lambda, 1e-10) * max(1, n)

    if (target_volume < min_volume - volume_tolerance or
            target_volume > max_volume + volume_tolerance):
        raise ValueError(
            f"目标体积 {target_volume:.6g} 不可行。给定掩码下可行总体积为 "
            f"[{min_volume:.6g}, {max_volume:.6g}]")

    # 对几乎可行的目标体积截断，避免舍入误差误报
    target_volume = min(max(target_volume, min_volume), max_volume)

    xnew = fixed_values.copy()
    if n_active == 0:
        return xnew, _build_info(
            np.nan, 0, xnew, target_volume, active_mask, True)

    # 灵敏度通常为负；设下限保护 sqrt() 并让这些单元向低密度更新
    dv = (np.ones_like(x) if opts["volume_sensitivity"] is None
          else np.asarray(opts["volume_sensitivity"], dtype=float))
    if dv.shape != x.shape or not np.all(np.isfinite(dv)) or np.any(dv < 0):
        raise ValueError("volume_sensitivity 必须为与 x 同形状的非负有限数组")
    dv = np.maximum(dv, 1e-12)
    sensitivity_ratio = np.maximum(1e-30, -dc[active_mask] / dv[active_mask])
    x_active = np.minimum(1.0, np.maximum(xmin, x[active_mask]))
    volume_fn = opts["volume_fn"] or (lambda value: float(np.asarray(value).mean()))

    l1, l2 = 0.0, 1e5
    bisect_iterations = 0
    converged = False

    for it in range(1, max_bisect + 1):
        bisect_iterations = it
        lam = 0.5 * (l1 + l2)
        candidate = x_active * np.sqrt(sensitivity_ratio / lam)
        candidate = np.maximum(
            xmin,
            np.maximum(x_active - move,
                       np.minimum(1.0, np.minimum(x_active + move, candidate))))
        xnew[active_mask] = candidate
        current_volume_fraction = float(volume_fn(xnew))
        if current_volume_fraction > volfrac:
            l1 = lam
        else:
            l2 = lam
        if (l2 - l1) <= tol_lambda:
            converged = True
            break

    # 用最终区间的中点重算密度，确保 lambda 与 xnew 对应
    lam = 0.5 * (l1 + l2)
    candidate = x_active * np.sqrt(sensitivity_ratio / lam)
    candidate = np.maximum(
        xmin,
        np.maximum(x_active - move,
                   np.minimum(1.0, np.minimum(x_active + move, candidate))))
    xnew[active_mask] = candidate
    xnew[passive_void] = xmin
    xnew[passive_solid] = 1.0

    info = _build_info(lam, bisect_iterations, xnew, target_volume,
                       active_mask, converged)
    info["physical_volume_fraction"] = float(volume_fn(xnew))
    info["physical_volume_error"] = info["physical_volume_fraction"] - volfrac
    return xnew, info


def _build_info(lam, bisect_iterations, xnew, target_volume,
                active_mask, converged) -> dict:
    return {
        "lambda": float(lam),
        "bisect_iterations": bisect_iterations,
        "volume": float(xnew.sum()),
        "volume_fraction": float(xnew.mean()),
        "target_volume": float(target_volume),
        "volume_error": float(xnew.sum() - target_volume),
        "active_elements": int(np.count_nonzero(active_mask)),
        "converged": bool(converged),
    }

File: matlab/engineering/TopOpt_2D/OC_solver.m
function [xnew, info] = OC_solver(x, dc, volfrac, opts)
%OC_SOLVER 面向柔顺度拓扑优化的增强 OC 密度更新器。
%   [XNEW, INFO] = OC_SOLVER(X, DC, VOLFRAC, OPTS) 根据滤波后的
%   柔顺度灵敏度 DC 更新密度场 X。未提供 OPTS 时，其更新规则与经典
%   99 行代码的 OC 方法一致；同时增加以下能力：
%     - 可配置的数值参数和单步密度变化上限；
%     - 被动空域 / 被动实体域；
%     - 体积约束的可行性检查；
%     - 通过 INFO 返回迭代诊断信息。
%
%   OPTS 的可选字段：
%     xmin          密度下限，默认 1e-3
%     move          每次迭代允许的最大密度变化，默认 0.2
%     tol_lambda    拉格朗日乘子的二分搜索精度，默认 1e-4
%     max_bisect    最大二分搜索次数，默认 100
%     passive_void  逻辑/数值掩码：对应单元始终为 xmin
%     passive_solid 逻辑/数值掩码：对应单元始终为 1
%     active_mask   逻辑/数值掩码：允许由 OC 更新的单元
%     volume_mask   体积分数统计区域；默认全部单元。L型设计域应传入
%                   与滤波相同的 domainMask，使 volfrac 相对有效区域定义
%
%   默认情况下 VOLFRAC 相对于 X 的全部单元定义，与 top.m 一致。
%   提供 volume_mask 后，目标材料量仅相对于该区域定义。

    if nargin < 4 || isempty(opts)
        opts = struct();
    end

    opts = set_default(opts, 'xmin',       1e-3);
    opts = set_default(opts, 'move',       0.2);
    opts = set_default(opts, 'tol_lambda', 1e-4);
    opts = set_default(opts, 'max_bisect', 100);
    opts = set_default(opts, 'passive_void',  []);
    opts = set_default(opts, 'passive_solid', []);
    opts = set_default(opts, 'active_mask',  []);
    opts = set_default(opts, 'volume_mask',  []);

    validateattributes(x, {'numeric'}, {'real', 'finite', 'nonempty'}, ...
        mfilename, 'x', 1);
    validateattributes(dc, {'numeric'}, {'real', 'finite', 'size', size(x)}, ...
        mfilename, 'dc', 2);
    validateattributes(volfrac, {'numeric'}, {'real', 'finite', 'scalar', ...
        '>=', 0, '<=', 1}, mfilename, 'volfrac', 3);
    validateattributes(opts.xmin, {'numeric'}, {'real', 'finite', 'scalar', ...
        '>', 0, '<', 1}, mfilename, 'opts.xmin');
    validateattributes(opts.move, {'numeric'}, {'real', 'finite', 'scalar', ...
        '>', 0}, mfilename, 'opts.move');
    validateattributes(opts.tol_lambda, {'numeric'}, {'real', 'finite', 'scalar', ...
        '>', 0}, mfilename, 'opts.tol_lambda');
    validateattributes(opts.max_bisect, {'numeric'}, {'real', 'finite', 'scalar', ...
        'integer', 'positive'}, mfilename, 'opts.max_bisect');

    n = numel(x);
    passive_void = make_mask(opts.passive_void, size(x), false);
    passive_solid = make_mask(opts.passive_solid, size(x), false);
    if any(passive_void(:) & passive_solid(:))
        error('OC_solver:OverlappingPassiveRegions', ...
            'passive_void and passive_solid must not overlap.');
    end

    if isempty(opts.active_mask)
        active_mask = ~(passive_void | passive_solid);
    else
        active_mask = make_mask(opts.active_mask, size(x), true);
        if any(active_mask(:) & (passive_void(:) | passive_solid(:)))
            error('OC_solver:ConflictingMasks', ...
                'An element cannot be active and passive at the same time.');
        end
    end

    volume_mask = make_mask(opts.volume_mask, size(x), true);
    if ~any(volume_mask(:))
        error('OC_solver:EmptyVolumeMask', ...
            'volume_mask must contain at least one element.');
    end
    if any(active_mask(:) & ~volume_mask(:))
        error('OC_solver:ActiveOutsideVolumeMask', ...
            'Every active element must also belong to volume_mask.');
    end

    % 未被显式标记为可设计的单元，固定在其当前密度。这样 active_mask
    % 可安全用于限定设计域，而不会意外改变域外单元。
    fixed_mask = ~active_mask;
    fixed_values = min(1, max(opts.xmin, x));
    fixed_values(passive_void) = opts.xmin;
    fixed_values(passive_solid) = 1.0;

    target_volume = volfrac * nnz(volume_mask);
    fixed_volume = sum(fixed_values(fixed_mask & volume_mask));
    n_active = nnz(active_mask);
    min_volume = fixed_volume + n_active * opts.xmin;
    max_volume = fixed_volume + n_active;
    volume_tolerance = max(opts.tol_lambda, 1e-10) * max(1, n);

    if target_volume < min_volume - volume_tolerance || ...
            target_volume > max_volume + volume_tolerance
        error('OC_solver:InfeasibleVolume', ...
            ['The requested volume %.6g is infeasible. Feasible total ', ...
             'volume is [%.6g, %.6g] with the supplied masks.'], ...
            target_volume, min_volume, max_volume);
    end

    % 对数值上几乎可行的目标体积进行截断，避免边界处仅因舍入误差或
    % 二分精度而误报“不可行”。
    target_volume = min(max(target_volume, min_volume), max_volume);

    xnew = fixed_values;
    if n_active == 0
        info = build_info(NaN, 0, xnew, target_volume, active_mask, ...
            volume_mask, true);
        return;
    end

    % 最小柔顺度问题的滤波灵敏度通常为负。若改进模型导致某些灵敏度
    % 为零或非负，设置下限可保护 sqrt()，并使这些单元向受步长限制的
    % 较低密度方向更新。
    sensitivity_ratio = max(1e-30, -dc(active_mask));
    x_active = min(1, max(opts.xmin, x(active_mask)));

    l1 = 0.0;
    l2 = 1e5;
    lambda = NaN;
    bisect_iterations = 0;
    converged = false;

    for iter = 1:opts.max_bisect
        bisect_iterations = iter;
        lambda = 0.5 * (l1 + l2);
        candidate = x_active .* sqrt(sensitivity_ratio ./ lambda);
        candidate = max(opts.xmin, max(x_active - opts.move, ...
                    min(1.0, min(x_active + opts.move, candidate))));

        xnew(active_mask) = candidate;
        current_volume = sum(xnew(volume_mask));

        if current_volume > target_volume
            l1 = lambda;
        else
            l2 = lambda;
        end

        if (l2 - l1) <= opts.tol_lambda
            converged = true;
            break;
        end
    end

    % 使用最终二分区间的中点重新计算密度，确保报告的 lambda 与 xnew
    % 相互对应；即使循环因达到 max_bisect 而退出也是如此。
    lambda = 0.5 * (l1 + l2);
    candidate = x_active .* sqrt(sensitivity_ratio ./ lambda);
    candidate = max(opts.xmin, max(x_active - opts.move, ...
                min(1.0, min(x_active + opts.move, candidate))));
    xnew(active_mask) = candidate;
    xnew(passive_void) = opts.xmin;
    xnew(passive_solid) = 1.0;

    info = build_info(lambda, bisect_iterations, xnew, target_volume, ...
        active_mask, volume_mask, converged);
end


function opts = set_default(opts, name, value)
    if ~isfield(opts, name) || isempty(opts.(name))
        opts.(name) = value;
    end
end


function mask = make_mask(value, expected_size, default_value)
    if isempty(value)
        mask = repmat(default_value, expected_size);
        return;
    end
    if ~isequal(size(value), expected_size)
        error('OC_solver:InvalidMaskSize', ...
            'All masks must have the same size as x.');
    end
    if ~islogical(value) && (~isnumeric(value) || any(~isfinite(value(:))))
        error('OC_solver:InvalidMask', ...
            'Masks must be logical or finite numeric arrays.');
    end
    mask = logical(value);
end


function info = build_info(lambda, bisect_iterations, xnew, target_volume, ...
        active_mask, volume_mask, converged)
    info.lambda = lambda;
    info.bisect_iterations = bisect_iterations;
    info.volume = sum(xnew(volume_mask));
    info.volume_fraction = mean(xnew(volume_mask));
    info.target_volume = target_volume;
    info.volume_error = info.volume - target_volume;
    info.active_elements = nnz(active_mask);
    info.volume_elements = nnz(volume_mask);
    info.converged = converged;
end
